File: backend/controller/item.js
const ItemDataAccess = require('../data-access/item');
const UserDataAccess = require('../data-access/user');
const puppeteer = require('puppeteer');
const bcrypt = require('bcryptjs');

module.exports.createItem = async (itemDetails) => {
	try {
		const user = itemDetails.user;
		const barcode = itemDetails.barcode;
		const name = itemDetails.name;

		const sameNameItem = await ItemDataAccess.getItem({ user, name });
		const sameBarcodeItem = await ItemDataAccess.getItem({ user, barcode });

		if (sameNameItem !== null || sameBarcodeItem !== null) {
			return { result: false }
		}

		const res = await ItemDataAccess.createItem(itemDetails);
		return { result: res }
	} catch (err) {
		console.error(err);
		return { result: false }
	}
}

module.exports.createItemWithoutBarcode = async (itemDetails) => {
	try {
		const user = itemDetails.user;
		const name = itemDetails.name;

		const sameNameItem = await ItemDataAccess.getItem({ user, name });

		if (sameNameItem !== null) {
			return { result: false }
		}

		const res = await ItemDataAccess.createItem(itemDetails);
		return { result: res }
	} catch (err) {
		console.error(err);
		return { result: false }
	}	
}

module.exports.getAllItems = async (userId) => {
	try {
		const items = await ItemDataAccess.getItems({ user: userId });
		return { result: items }
	} catch (err) {
		console.error(err);
		return { result: false }
	}
	
}

module.exports.editItemNoPassword = async ({ userId, itemId, edits }) => {
	try {
		const query = { user: userId, _id: itemId};
		const res = await ItemDataAccess.editItem(query, edits);
		return { result: res }
	} catch (err) {
		console.error(err);
		return { result: false }
	}
}

module.exports.editItem = async ({userId, itemId, edits, password}) => {
	try {
		const query = { user: userId, _id: itemId};

		const user = await UserDataAccess.getUser({ _id: userId });

		if (!!user.itemPassword) {
			const matchedPasswords = 
				bcrypt.compareSync(password, user.itemPassword);
			if (!matchedPasswords) return { result: false }
		}

		const res = await ItemDataAccess.editItem(query, edits);
		return { result: res }
	} catch (err) {
		console.error(err);
		return { result: false }
	}
}

module.exports.deleteItem = async ({ userId, itemId }) => {
	try {
		const query = { user: userId, _id: itemId };
		const res = await ItemDataAccess.deleteItem(query);
		return { result: res }
	} catch (err) {
		console.error(err);
		return { result: false }
	}
}

function addLeadingZero(num) {
	return (num.toString().length === 1) ? `0${num}` : `${num}`;
}

module.exports.getInventoryCSV = async(userId) => {
	try {
		const items = await ItemDataAccess.getItems({ user: userId });
		let date = new Date();
		const dateString = `${date.getFullYear()}${addLeadingZero(date.getMonth() + 1)}` +
			`${addLeadingZero(date.getDate())}`
		let csvString = `Product Name,Stock,Price, Expiry Date\n`
		items.forEach(item => csvString +=
`${item.name},${item.stock},${item.price},${formatExpiryDate(item.expiryDates[0])}\n`);
		return { result: { csvString, date: dateString } }
	} catch(err) {
		console.error(err);
		return { result: false }
	}
}

function formatExpiryDate(unix) {
  if (unix) {
    return (new Date(unix)).toLocaleDateString();
  } else {
    return '';
  }
}

module.exports.getInvalidBarcodes = async (userId) => {
	try {
		const query = {
		user: userId,
			barcode: {
				$ne: null
			},
			isValidBarcode: {
				$ne: true
			}
		}
		const items = await ItemDataAccess.getItems(query);
		const invalidBarcodes = [];
		const browser = await puppeteer.launch({
			headless: true,
			args: [
				'--no-sandbox',
				'--disable-setuid-sandbox'
			]
		});
		const page = await browser.newPage();

		for (let i = 0; i < items.length; i++) {
			const barcode = items[i].barcode;
			const eanLength = 13;
			const upcLength = 12;
			if (barcode.length > eanLength || barcode.length < upcLength ) {
				invalidBarcodes.push(items[i]);
				continue;
			}

			if (items[i].isValidBarcode === false && items[i].isBarcodeChecked === true) {
				invalidBarcodes.push(items[i]);
				continue;
			}

			await page.goto(`https://www.google.com/search?q=barcode+${barcode}`);
			const hasResults = await page.evaluate(() => {
				if (document.querySelector('#result-stats') === null) {
					return false;
				}
				return true
			});

			const editQuery = {user: userId, _id: items[i]._id}
			let edits;

			if (hasResults === true) {
				edits = {isValidBarcode: true, isBarcodeChecked: true}
			} else if (hasResults === false) {
				edits = {isBarcodeChecked: true} 
				invalidBarcodes.push(items[i])
			}

			await ItemDataAccess.editItem(editQuery, edits);
		}


		return { result: invalidBarcodes }
	} catch (err) {
		console.error(err);
		return { result: false }
	}
}

File: backend-go/src/handlers/item.go
package handlers

import (
	"fmt"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
	"golang.org/x/crypto/bcrypt"

	"bc-pos-backend/internal/middleware"
	"bc-pos-backend/internal/models"
)

// POST /api/items  — create item with barcode (checks name + barcode uniqueness per user)
func CreateItem(c *gin.Context) {
	claims := c.MustGet(middleware.ClaimsKey).(*middleware.Claims)
	userID := claims.ID

	var item models.Item
	if err := c.ShouldBindJSON(&item); err != nil {
		c.JSON(200, gin.H{"result": false})
		return
	}

	if exists, _ := dataStore.ItemExistsByName(userID, item.Name); exists {
		c.JSON(200, gin.H{"result": false})
		return
	}
	if item.Barcode != "" {
		if exists, _ := dataStore.ItemExistsByBarcode(userID, item.Barcode); exists {
			c.JSON(200, gin.H{"result": false})
			return
		}
	}

	if item.ExpiryDates == nil {
		item.ExpiryDates = []int64{}
	}

	if err := dataStore.CreateItem(userID, item); err != nil {
		c.JSON(200, gin.H{"result": false})
		return
	}

	c.JSON(200, gin.H{"result": item})
}

// POST /api/items/no-barcode  — create item without barcode (checks name uniqueness per user)
func CreateItemWithoutBarcode(c *gin.Context) {
	claims := c.MustGet(middleware.ClaimsKey).(*middleware.Claims)
	userID := claims.ID

	var item models.Item
	if err := c.ShouldBindJSON(&item); err != nil {
		c.JSON(200, gin.H{"result": false})
		return
	}

	if exists, _ := dataStore.ItemExistsByName(userID, item.Name); exists {
		c.JSON(200, gin.H{"result": false})
		return
	}

	if item.ExpiryDates == nil {
		item.ExpiryDates = []int64{}
	}

	if err := dataStore.CreateItem(userID, item); err != nil {
		c.JSON(200, gin.H{"result": false})
		return
	}

	c.JSON(200, gin.H{"result": item})
}

// GET /api/items/all  — return all items belonging to the authenticated user
func GetAllItems(c *gin.Context) {
	claims := c.MustGet(middleware.ClaimsKey).(*middleware.Claims)
	items, err := dataStore.ListItemsByUser(claims.ID)
	if err != nil {
		c.JSON(200, gin.H{"result": false})
		return
	}
	c.JSON(200, gin.H{"result": items})
}

// PATCH /api/items/:itemId  — edit item; verifies itemPassword if one is set
func EditItem(c *gin.Context) {
	claims := c.MustGet(middleware.ClaimsKey).(*middleware.Claims)
	userID := claims.ID
	itemID := c.Param("itemId")

	var body struct {
		Edits    map[string]interface{} `json:"edits"`
		Password string                 `json:"password"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		c.JSON(200, gin.H{"result": false})
		return
	}

	user, err := dataStore.FindUserByID(userID)
	if err != nil || user == nil {
		c.JSON(200, gin.H{"result": false})
		return
	}

	if user.ItemPassword != "" {
		if err := bcrypt.CompareHashAndPassword([]byte(user.ItemPassword), []byte(body.Password)); err != nil {
			c.JSON(200, gin.H{"result": false})
			return
		}
	}

	if err := dataStore.UpdateItemByIDAndUser(userID, itemID, body.Edits); err != nil {
		c.JSON(200, gin.H{"result": false})
		return
	}

	c.JSON(200, gin.H{"result": true})
}

// DELETE /api/items/:itemId  — delete an item belonging to the authenticated user
func DeleteItem(c *gin.Context) {
	claims := c.MustGet(middleware.ClaimsKey).(*middleware.Claims)
	if err := dataStore.DeleteItemByIDAndUser(claims.ID, c.Param("itemId")); err != nil {
		c.JSON(200, gin.H{"result": false})
		return
	}
	c.JSON(200, gin.H{"result": true})
}

// GET /api/items/csv-inventory  — returns a CSV string of the user's inventory
func GetInventoryCSV(c *gin.Context) {
	claims := c.MustGet(middleware.ClaimsKey).(*middleware.Claims)
	items, err := dataStore.ListItemsByUser(claims.ID)
	if err != nil {
		c.JSON(200, gin.H{"result": false})
		return
	}

	now := time.Now()
	dateString := fmt.Sprintf("%04d%02d%02d", now.Year(), now.Month(), now.Day())

	var sb strings.Builder
	sb.WriteString("Product Name,Stock,Price, Expiry Date\n")
	for _, item := range items {
		sb.WriteString(fmt.Sprintf("%s,%v,%v,%s\n", item.Name, item.Stock, item.Price, formatExpiryDate(item.ExpiryDates)))
	}

	c.JSON(200, gin.H{"result": gin.H{"csvString": sb.String(), "date": dateString}})
}

// GET /api/items/invalid-barcodes  — returns items with invalid or flagged barcodes.
func GetInvalidBarcodes(c *gin.Context) {
	claims := c.MustGet(middleware.ClaimsKey).(*middleware.Claims)
	items, err := dataStore.ListInvalidBarcodeCandidates(claims.ID)
	if err != nil {
		c.JSON(200, gin.H{"result": false})
		return
	}

	const eanLen = 13
	const upcLen = 12

	invalid := make([]models.Item, 0)
	for _, item := range items {
		barcode := item.Barcode
		if len(barcode) > eanLen || len(barcode) < upcLen {
			invalid = append(invalid, item)
			continue
		}
		if !item.IsValidBarcode && item.IsBarcodeChecked {
			invalid = append(invalid, item)
		}
	}

	c.JSON(200, gin.H{"result": invalid})
}

// editItemStock increments/decrements an item's stock without a password check.
func editItemStock(userID, itemID string, delta float64) error {
	return dataStore.IncrementItemStock(userID, itemID, delta)
}

func formatExpiryDate(dates []int64) string {
	if len(dates) == 0 || dates[0] == 0 {
		return ""
	}
	t := time.Unix(dates[0]/1000, 0)
	return fmt.Sprintf("%d/%d/%d", t.Month(), t.Day(), t.Year())
}
